File: Search/permutationIndexII/permutationIndex.py
class Solution:
    def permutationIndex(self, num):
        if num is None or len(num) == 0:
            return None
            
        index = 0
        factor = 0
        for i in xrange(len(num) - 1, -1, -1):
            rank = 0
            for j in xrange(i + 1, len(num)):
                if num[i] > num[j]:
                    rank += 1
                    
            index += rank * factor
            factor *= (len(num) - i)
            
        return index  
                
                

File: Search/permutationIndexII/permutationIndex.cc
class Solution {
    public:
        long long permutationIndex(vector<int> &nums) {
            if (nums == NULL || nums.size() == 0)
                return;
            
            long long index = 0;
            long long factor = 0;
            for(int i = nums.size() - 1; i >= 0; --i) {
                int rank = 0;
                for(int j = i; j < nums.size(); ++j) {
                    if (nums[i] > nums[j]) ++rank;
                }
                index += rank * factor;
                factor *= (nums.size() - i);
                    
            }
            return index;
            
        }
};